server/athenian/api/internal/team_sync.py

```diff
@@ -0,0 +1,325 @@
+from athenian.api.db import Connection, Database, DatabaseLike, Row, is_postgresql
+from athenian.api.internal.team import (
+    get_meta_teams_members,
+    get_meta_teams_topological_order,
+    get_root_team,
+)
+from athenian.api.models.metadata.github import (
+    Organization as MetadataOrganization,
+    Team as MetadataTeam,
+)
+from athenian.api.models.state.models import Team
+
+log = logging.getLogger(__name__)
+
+
+async def sync_teams(
+    account: int,
+    meta_ids: Sequence[int],
+    sdb: Database,
+    mdb: DatabaseLike,
+    *,
+    dry_run: bool = False,
+) -> None:
+    """Sync the teams from metadata DB to state DB."""
+    meta_teams = await _MetaTeams.from_db(meta_ids, mdb)
+    all_members = await meta_teams.get_all_members()
+    log.info(
+        "executing team sync for account %s%s", account, " in dry run mode" if dry_run else "",
+    )
+    async with sdb.connection() as sdb_conn:
+        async with sdb_conn.transaction():
+            state_teams = await _StateTeams.from_db(account, sdb_conn)
+
+            new_meta_teams = meta_teams.get_new(state_teams)
+            existing_meta_teams = meta_teams.get_existing(state_teams)
+            gone_state_teams = state_teams.get_gone(meta_teams)
+
+            db_ops = _DryRunStateDBOperator() if dry_run else _StateDBOperator(account, sdb_conn)
+            await db_ops.insert_new(new_meta_teams, all_members, state_teams)
+            await db_ops.update_existing(existing_meta_teams, all_members, state_teams)
+            await db_ops.delete(gone_state_teams)
+            await db_ops.revert_temporary_names()
+
+
+class SyncTeamsError(Exception):
+    """An error occurred during teams sync operation."""
+
+
+class _MetaTeams:
+    def __init__(self, rows: Sequence[Row], meta_ids: Sequence[int], mdb: DatabaseLike):
+        self._rows = rows
+        self._meta_ids = meta_ids
+        self._mdb = mdb
+        self._by_id = {r[MetadataTeam.id.name]: r for r in rows}
+
+    @classmethod
+    async def from_db(cls, meta_ids: Sequence[int], mdb: DatabaseLike) -> _MetaTeams:
+        meta_orgs_query = sa.select(MetadataOrganization.id).where(
+            MetadataOrganization.acc_id.in_(meta_ids),
+        )
+        columns = (MetadataTeam.name, MetadataTeam.id, MetadataTeam.parent_team_id)
+        query = sa.select(columns).where(
+            MetadataTeam.acc_id.in_(meta_ids), MetadataTeam.organization_id.in_(meta_orgs_query),
+        )
+        return cls(await mdb.fetch_all(query), meta_ids, mdb)
+
+    def get_new(self, state_teams: _StateTeams) -> Sequence[Row]:
+        existing = state_teams.existing_by_origin_node_id
+        return [r for r in self._rows if r[MetadataTeam.id.name] not in existing]
+
+    def get_existing(self, state_teams: _StateTeams) -> Sequence[Row]:
+        existing = state_teams.existing_by_origin_node_id
+        return [r for r in self._rows if r[MetadataTeam.id.name] in existing]
```

Comment:
Performance: instead of iterating twice, define the method `triage() -> tuple[list[Row], list[Row]]` that splits the rows into new and existing. We don't use `get_new()` and `get_existing()` separately.